### PR TITLE
Metal: Use setBytes:length:atIndex for arguments

### PR DESCRIPTION
**Pull request by shoaibkamil**
Instead of allocating a temporary `MTLBuffer`, if the arguments
size is less than 4KB, use `setBytes:length:atIndex`.  
See https://developer.apple.com/reference/metal/mtlcomputecommandencoder/1443159-setbytes?language=objc#discussion for more information.

This change should never be slower than the existing approach, and helps remove one source of overhead for small pipelines.

CC @estollnitz

### CODE REVIEW

File: src/runtime/metal.cpp [hunk 1]
```diff
@@ -671,14 +679,29 @@ WEAK int halide_metal_run(void *user_context,
...
-            error(user_context) << "Metal: Could not allocate arguments buffer.\n";
-            release_ns_object(pipeline_state);
-            release_ns_object(function);
-            return -1;
+        mtl_buffer *args_buffer = 0;        // used if the total arguments size large
+        uint8_t *small_args_buffer = 0;     // used if the total arguments size is small
+        char *args_ptr;
+
+        if (total_args_size < 4096) {
+            small_args_buffer = (uint8_t*)malloc(sizeof(uint8_t) * total_args_size);
```
Comment (abadams) [inline src/runtime/metal.cpp]: Is it worth just using 4K of stack here?

**Comment by zvookin** [comment]
The docs seem to indicate "setBytes:length:atIndex:" is iOS 8.3 and later, not iOS 8.0 and later which is the baseline for Metal. Idea was to run on anything that supports Metal unless there's a strong reason not to. Anyone more in the know able to comment on this? If it is indeed not supported and offers performance wins of note on new enough versions of Metal, we cna either case on the Metal version or check if the selector is supported.

File: src/runtime/metal.cpp (hunk 1 above)
Comment (shoaibkamil) [inline src/runtime/metal.cpp]: Probably; I'll make that change.

**Comment by shoaibkamil** [comment]
@zvookin You're right, this appears to be iOS 8.3+.  I can implement either a selector check or version check; please let me know which you prefer.

Re: performance wins, we have an internal version of the Metal runtime with several improvements for running multiple small pipelines in a row.  This change gives us ~30% performance improvement, though the improvement is less without our other changes, for which I'll open separate PRs.

**Comment by zvookin** [comment]
I generally prefer selector checks because they tie the choice of which
code to use to a specific property of the API. The downside is as there
gets to be many cases, each one requires a selector test where the version
is only requested once. Though that gets into whether the decision is
cached somehow.  I expect you or someone working on this has more expertise
than I do.

Very glad to see this backend getting further improvements!

-Z-

**Comment by shoaibkamil** [comment]
Sorry for the delay on this.  I've rebased and added a selector check that is cached (and invalidated if the device pointer changes).  I've also addressed @abadams's comment and made the temporary buffer live on the stack.

File: src/runtime/metal.cpp [hunk 2]
```diff
@@ -195,6 +210,11 @@ struct module_state {
 };
 WEAK module_state *state_list = NULL;
 
+// API Capabilities.  If more capabilities need to be checked,
+// this can be refactored to something more robust/general.
+bool halide_metal_api_supports_set_bytes;
```
Comment (zvookin) [inline src/runtime/metal.cpp]: These need to be WEAK and should not be prefixed with "halide_" as they are not exported. I'm guessing that if they are not WEAK, the runtime linker will insist that they are prefixed with "halide_". Is there a desire to export the variables or did you run into an issue with this? I can just merge it and make the fix afterward if that is expedient, but figured I'd check in re: intent and history first.

Comment (shoaibkamil) [inline src/runtime/metal.cpp]: No, you're right-- they should be WEAK, and they should not be exported.  I'll change this & push shortly.

**Comment by zvookin** [comment]
Apologies for the delay. Was surprised to find this had not been merged yet.

**Pull request merged**